### PR TITLE
planner: support the Group and GroupExpr for the cascades planner

### PR DESCRIPTION
**Pull request by zz-jason**
### What problem does this PR solve? 

split from https://github.com/pingcap/tidb/pull/7869

Add the Group/GroupExpr support

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

### CODE REVIEW

File: planner/cascades/group_test.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cascades
+
+import (
+	// "container/list"
```
Comment (XuHuaiyu) [inline planner/cascades/group_test.go]: remove these two lines?

**Comment by zz-jason** [comment]
@XuHuaiyu done, PTAL

File: planner/cascades/group.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+// FingerPrint returns the unique fingerprint of the group.
+func (g *Group) FingerPrint() string {
+	if g.selfFingerprint == "" {
+		g.selfFingerprint = fmt.Sprintf("%p", g)
+	}
+	return g.selfFingerprint
+}
+
+// Insert a nonexistent group exxpression.
```
Comment (lamxTyler) [inline planner/cascades/group.go]: `expression`.

**Comment by zz-jason** [comment]
@XuHuaiyu @lamxTyler PTAL

**Pull request merged**